pkg/eventcache/eventcache.go

```diff
@@ -53,81 +60,183 @@ var (
+//     reference count.
+// (b) EventCache is responsible to handle OOO events that miss Process, Parent,
+//     or their PodInformation. As handling reference counting is compicated
+//     when we have OOO events the next functions simplify the iteraction with
+//     the EventCache.
+
+// This function adds process/parent information to the event e. This does not
+// affect refcnt. This is ideal for events like kprobe/tracepoint that do not
+// come into pairs. This transparently returns the event in the case we don't
+// miss anything or adds the event to the EventCache. In that case it uses processLRU
+// to get process and parent info.
+func AddProcessParent(e notify.Event, msg notify.Message, pid uint32, ktime uint64, timestamp uint64) *tetragon.GetEventsResponse {
+	f := func(*process.ProcessInternal) {}
+	return addProcInfo(e, msg, pid, ktime, timestamp, f)
+}
+
+// The AddProcessParentRef{Inc, Dec} functions do the similar to AddProcessParent
+// with the only difference to also incread/decrease the recnt of ProcessInternal.
+// These should be used in pairs.
+//
+// As for now they used in exec/clone/exit events. Exec/clone events create a new
+// ProcessInternal with refcnt equals to 1 and increases the reference count of the
+// parent. Exit events descreases the reference count of the process and parent.
+//
+// Other examples can be on file events (i.e. increase on file open and decrease on
+// file close) or networking events (i.e. increase on connect/accept/listen and
+// decrease on close).
+func AddProcessParentRefInc(e notify.Event, msg notify.Message, pid uint32, ktime uint64, timestamp uint64) *tetragon.GetEventsResponse {
+	f := func(p *process.ProcessInternal) { p.RefInc() }
+	return addProcInfo(e, msg, pid, ktime, timestamp, f)
+}
+
+func AddProcessParentRefDec(e notify.Event, msg notify.Message, pid uint32, ktime uint64, timestamp uint64) *tetragon.GetEventsResponse {
+	f := func(p *process.ProcessInternal) { p.RefDec() }
+	return addProcInfo(e, msg, pid, ktime, timestamp, f)
+}
 
+func addProcInfo(e notify.Event, msg notify.Message, pid uint32, ktime uint64, timestamp uint64, pFn procFn) *tetragon.GetEventsResponse {
+	var tetragonParent, tetragonProcess *tetragon.Process
+
+	proc, parent := process.GetParentProcessInternal(pid, ktime)
+	if proc != nil {
+		pFn(proc)
+		tetragonProcess = proc.GetProcessCopy()
+		e.SetProcess(tetragonProcess)
+	}
 	if parent != nil {
-		ev.SetParent(parent.GetProcessCopy())
-	} else {
-		errormetrics.ErrorTotalInc(errormetrics.EventCacheParentInfoFailed)
-		err = ErrFailedToGetParentInfo
+		pFn(parent)
+		tetragonParent = parent.GetProcessCopy()
+		e.SetParent(tetragonParent)
+	}
+
+	act := &notify.CacheActions{
+		NeedProcess:    NeededProcess(tetragonProcess),
+		NeedProcessPod: NeededPod(tetragonProcess),
+		NeedParent:     pid > 1 && NeededProcess(tetragonParent),
+		NeedParentPod:  pid > 1 && NeededPod(tetragonParent),
 	}
```

Comment:
How about we call the pFn only if NeedProcess is not set instead of if process is not nill?
pFn is called in the retry if `NeedProcess` or `NeedParent` is set.

But:
```
func NeededProcess(proc *tetragon.Process) bool {
	if proc == nil {
		return true
	}
	if proc.Binary == "" {
		return true
	}
	return false
}
```

so if  p != nil but p.Binary == "", I think we would call pFn twice. Once here and once whenever things are setup in the cache.

The same thing holds for the parent.

Reply:
I see what you are saying there. I don't believe that we can have an entry in the cache (i.e. `proc != nil`) but it will be not full (i.e. `proc.Binary == ""`). 

I will change that to what you propose just to be sure that I don't miss any corner cases.

Reply:
> I see what you are saying there. I don't believe that we can have an entry in the cache (i.e. proc != nil) but it will be not full (i.e. proc.Binary == "").

Not sure I understand.

Are you saying that there is no way to to have an entry in the cache with Binary == ""?

Reply:
Yes, I believe that this cannot happen but I want to check that again.